app/javascript/src/Policies/components/CustomPolicy.jsx

```diff
@@ -0,0 +1,143 @@
+import type {InputEvent} from 'Policies/types'
+import 'codemirror/mode/javascript/javascript'
+import 'Policies/styles/policies.scss'
+
+type OnChange = (InputEvent) => void
+
+const CM_OPTIONS = {
+  theme: 'default',
+  height: 'auto',
+  viewportMargin: Infinity,
+  mode: {
+    name: 'javascript',
+    json: true,
+    statementIndent: 2
+  },
+  lineNumbers: true,
+  lineWrapping: true,
+  indentWithTabs: false,
+  tabSize: 2
+}
+
+const POLICY_TEMPLATE: Policy = {
+  schema: { configuration: {} },
+  directory: '',
+  id: 0
+}
+
+function CSRFToken ({win = window}: {win?: any}): React.Node {
+  const getMetaContent = meta => win.document.head.querySelector(`meta[name~=${meta}][content]`).content
+  const props = {
+    name: getMetaContent('csrf-param'),
+    value: getMetaContent('csrf-token'),
+    type: 'hidden'
+  }
+  return (
+    <input {...props} />
+  )
+}
+
+function Editor ({onChange, code}: {onChange: (Object) => void, code: Object}): React.Node {
+  const [ state, setState ] = useState({valid: true, code: toJson(code)})
+
+  const getConfiguration = (schema) => {
+    return (schema.configuration)
+      ? schema.configuration
+      : (function () { throw new Error('Policy configuration not found') })()
+  }
+
+  const onCodeChange = (editor, metadata, code) => {
+    try {
+      onChange(getConfiguration(fromJson(code)))
+      setState({ valid: true, code })
+    } catch (err) {
+      console.warn(err)
+      setState({ valid: false, code })
+    }
+  }
+
+  const [ icon, cls, hiddenClass, errorClass ] = state.valid
+    ? [ 'check', 'valid', 'is-hidden', '' ]
+    : [ 'times', 'invalid', '', 'CustomPolicy-error' ]
+
+  return (
+    <div className={`${errorClass} panel panel-default`}>
+      <div className="panel-heading">
+        <i className={`${cls} fa fa-${icon}`} />
+        {' JSON Schema'}
+      </div>
+      <CodeMirror
+        value={state.code}
+        onChange={onCodeChange}
+        autoCursor={false}
+        options={CM_OPTIONS}
+      />
+      <div className={hiddenClass}> There's an error in the JSON Schema</div>
+    </div>
+  )
+}
+
+function CustomPolicyForm ({policy, onChange}: {policy: Policy, onChange: OnChange}): React.Node {
+  const isNewPolicy = !(policy.id && policy.id !== 0)
+  const action = (isNewPolicy) ? '/p/admin/registry/policies/' : `/p/admin/registry/policies/${policy.id}`
+  return (
+    <form action={action} method="post">
+      <label>
+        Directory:
+        <input type="text" name="directory" value={policy.directory} onChange={onChange} disabled={!isNewPolicy} />
+      </label>
+      <input type="hidden" name="id" value={policy.id} disabled={true} />
+      <input name="schema" type="hidden" value={JSON.stringify(policy.schema)} />
+      {(!isNewPolicy) ? <input name="_method" type="hidden" value='put' /> : ''}
+      <input type="submit" />
+      <CSRFToken />
+    </form>
+  )
+}
+
+function Form ({policy}: {policy: Policy}): React.Node {
```

Comment:
Comment was outdated. What about naming this `defaultPolicy` or `initialPolicy` and then use `policy` to track the state? Shortened var names should be avoided IMO.

Reply:
The form expects a Policy, a policy is what it gets. It's fed by Rails, so the only way it could match a "initial" or "default" would be in the "new" policy case. ... Or maybe I'm missing something?

Reply:
I'm not talking about the business logic, just naming convention. Having `policy`, `setPolicy` and `pol` is confusing and misleading: it's not clear whether `setPolicy` is to be used with one or the other.

This is the right way `const [policy, setPolicy = useState(initialPolicy);` (See [#useState](https://reactjs.org/docs/hooks-reference.html#usestate)).

_initialPolicy_ only means it's the data defining the component's initial state, it doesn't matter where it came from or what it is (from a React pov).

Reply:
Makes sense